### ISSUE TITLE
How to build eos at Mac system

### ISSUE THREAD

**Issue by darkstarslyp** [patrikhuber/eos#163]
I want build eos at mac system, can someone help me .

**Comment by patrikhuber** [comment]
I don't personally have a mac, but I know from users that it compiles fine on mac :-) You're probably be off with installing cmake from homebrew, and opencv and boost as well, in case you need it for the example apps.

**Comment by darkstarslyp** [comment]
I have successed   build eos at Mac system ,so I will close this question.

**Comment by patrikhuber** [comment]
Yea, so, from v0.13.0 and later, eos doesn't build out-of-the-box on macOS because even the latest Xcode doesn't have `std::optional` yet. It has `std::experimental::optional` but it's a non-standard conforming version.
In any case, I'll bring back official macOS/Xcode support in the next couple of weeks or so.

**Comment by jkuri** [comment]
I was able to compile it on MacOS from current `master` branch (fc9156d).

First, I installed latest version of clang available on brew (5.0.1)
```sh
brew install --with-toolchain llvm
```
then I specified installed compiler with cmake flags
```sh
mkdir -p build && cd build
cmake -D CMAKE_C_COMPILER=/usr/local/opt/llvm/bin/clang -D CMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ ..
make -j4
```
got some nasty warnings from `rapidjson` which is included in `cereal` but it works.

Note that I also tried with `g++-7` but encountered nothing but problems.
I think the most cross-compilation solution would be using `boost/optional.hpp`, but not sure what does @patrikhuber thinks about that.
Hope this helps someone.

**Comment by patrikhuber** [comment]
Hi @jkuri,
Thank you very much for detailing the steps!
Indeed eos should compile with the default AppleClang in Xcode 9.2, but it currently doesn't, because of `optional`. I definitely do not want `boost/optional` but I plan to add proper support for Xcode (with `akrzemi1/Optional`) in the next few weeks.

**Comment by patrikhuber** [comment]
eos can now transparently use akrzemi1/Optional (it's included in the eos headers) when there is no `<optional>` yet. Currently there is a check for `#ifdef __APPLE__`. So eos now compiles fine on Xcode 9.2 and it's also CI-tested on travis macOS.
The changes are merged into `master` and present from eos [v0.15.0](https://github.com/patrikhuber/eos/releases/tag/v0.15.0).
